File: joss/comparisons/cat-2-type-safety/22-temporal-type-safety/example-temporal-type-safety.py
"""
Error Class 22: Temporal Type Safety

Python's pd.to_datetime silently produces NaT for invalid dates.
Date arithmetic with timedelta is allowed but pd.Timestamp + int
is silently deprecated or produces unexpected results.
"""
import pandas as pd

encounters = pd.DataFrame({
    "patient_id": ["P001", "P002", "P003"],
    "admit_date": ["2024-01-15", "2024-02-20", "not-a-date"],
    "los_days": [3, 7, 5],
})

# SILENT: Invalid date string becomes NaT
encounters["admit_date"] = pd.to_datetime(encounters["admit_date"], errors="coerce")
print(encounters["admit_date"])  # P003: NaT — no error, no warning

# SILENT: Arithmetic on NaT propagates
encounters["discharge"] = encounters["admit_date"] + pd.to_timedelta(encounters["los_days"], unit="D")
print(encounters["discharge"])  # P003: NaT — silently wrong

File: joss/comparisons/cat-2-type-safety/22-temporal-type-safety/example-temporal-type-safety.R
# Error Class 22: Temporal Type Safety
#
# R's as.Date produces NA for invalid dates with only a warning.
# Arithmetic on NA dates propagates silently.

library(tidyverse)

encounters <- tibble(
  patient_id = c("P001", "P002", "P003"),
  admit_date = c("2024-01-15", "2024-02-20", "not-a-date"),
  los_days = c(3, 7, 5)
)

# WARNING: Invalid date becomes NA
encounters <- encounters %>%
  mutate(admit_date = as.Date(admit_date))
# Warning: NAs introduced by coercion — but execution continues

# SILENT: Arithmetic on NA date propagates
encounters <- encounters %>%
  mutate(discharge = admit_date + los_days)
print(encounters$discharge)  # P003: NA — silently wrong
